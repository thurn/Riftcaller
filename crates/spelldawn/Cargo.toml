[package]
name = "spelldawn"
version = "0.0.0"
homepage = "https://github.com/thurn/spelldawn"
description = "Spelldawn: An asymmetric trading card game"
license = "Apache-2.0"
default-run = "spelldawn"
edition = "2021"

[[bin]]
name = "spelldawn"
path = "src/main.rs"
test = false
doctest = false

[dependencies]
cards_all = { path = "../cards/cards_all", version = "0.0.0" }
protos = { path = "../protos", version = "0.0.0" }
server = { path = "../server", version = "0.0.0" }

anyhow = "1.0.58"
tokio = { version = "1.19.2", features = ["macros", "rt-multi-thread", "sync", "time"] }
tonic = { version = "0.7.2", features = ["compression"] }
tonic-web = "0.3.0"
tracing = "0.1.35"
tracing-subscriber = "0.3.14"

[dev-dependencies]
rusty-hook = "0.11.2"
