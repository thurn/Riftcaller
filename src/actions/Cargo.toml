[package]
name = "actions"
version = "0.0.0"
description = "Top-level routing for user actions"
license = "Apache-2.0"
edition = "2021"

[lib]
path = "src/actions.rs"
doctest = false
test = false

[dependencies]
anyhow = "1.0.69"
tracing = "0.1.37"
enum-iterator = "1.3.0"

game_data = { path = "../data/game_data", version = "0.0.0" }
raids = { path = "../raids", version = "0.0.0" }
rules = { path = "../rules", version = "0.0.0" }
with_error = { path = "../with_error", version = "0.0.0" }
