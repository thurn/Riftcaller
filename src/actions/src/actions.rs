// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Contains functions for responding to user-initiated game actions received
//! from the client. The `handle_user_action` function is the primary
//! entry-point into the rules engine.

pub mod legal_actions;

use anyhow::Result;
use game_data::card_definition::AbilityType;
use game_data::card_state::CardPosition;
use game_data::delegates::{
    AbilityActivated, ActivateAbilityEvent, CardPlayed, CastCardEvent, DrawCardActionEvent,
};
use game_data::game::{GamePhase, GameState, HistoryEntry, HistoryEvent, MulliganDecision};
use game_data::game_actions::{CardTarget, GameAction, PromptAction};
use game_data::primitives::{AbilityId, CardId, RoomId, Side};
use game_data::updates::{GameUpdate, InitiatedBy};
use raids::RaidDataExt;
use rules::mana::ManaPurpose;
use rules::{card_prompt, dispatch, flags, mana, mutations, queries};
use tracing::{debug, instrument};
use with_error::{fail, verify, WithError};

/// Top level dispatch function responsible for mutating [GameState] in response
/// to all [GameAction]s
pub fn handle_game_action(
    game: &mut GameState,
    user_side: Side,
    action: &GameAction,
) -> Result<()> {
    match action {
        GameAction::PromptAction(prompt_action) => {
            handle_prompt_action(game, user_side, *prompt_action)
        }
        GameAction::Resign => handle_resign_action(game, user_side),
        GameAction::GainMana => gain_mana_action(game, user_side),
        GameAction::DrawCard => draw_card_action(game, user_side),
        GameAction::PlayCard(card_id, target) => {
            play_card_action(game, user_side, *card_id, *target)
        }
        GameAction::ActivateAbility(ability_id, target) => {
            activate_ability_action(game, user_side, *ability_id, *target)
        }
        GameAction::UnveilCard(card_id) => unveil_action(game, user_side, *card_id),
        GameAction::InitiateRaid(room_id) => {
            raids::handle_initiate_action(game, user_side, *room_id)
        }
        GameAction::LevelUpRoom(room_id) => level_up_room_action(game, user_side, *room_id),
        GameAction::SpendActionPoint => spend_action_point_action(game, user_side),
    }?;

    Ok(())
}

/// Returns true if the indicated player currently has a legal game action
/// available to them.
pub fn has_priority(game: &GameState, side: Side) -> bool {
    match &game.info.phase {
        GamePhase::ResolveMulligans(_) => return flags::can_make_mulligan_decision(game, side),
        GamePhase::GameOver { .. } => return false,
        _ => {}
    };

    match &game.info.raid {
        Some(raid) => side == raid.phase().active_side(),
        None => side == game.info.turn.side,
    }
}

fn handle_resign_action(game: &mut GameState, side: Side) -> Result<()> {
    debug!(?side, "Applying resign action");
    if !matches!(game.info.phase, GamePhase::GameOver { .. }) {
        mutations::game_over(game, side.opponent())?;
    }
    Ok(())
}

/// Handles a choice to keep or mulligan an opening hand
fn handle_mulligan_decision(
    game: &mut GameState,
    user_side: Side,
    decision: MulliganDecision,
) -> Result<()> {
    verify!(
        flags::can_make_mulligan_decision(game, user_side),
        "Cannot make mulligan decision for {:?}",
        user_side
    );

    debug!(?user_side, ?decision, "Applying mulligan action");
    let mut mulligans = match &mut game.info.phase {
        GamePhase::ResolveMulligans(mulligans) => mulligans,
        _ => fail!("Incorrect game phase"),
    };

    match user_side {
        Side::Overlord => mulligans.overlord = Some(decision),
        Side::Champion => mulligans.champion = Some(decision),
    }

    let hand = game.card_list_for_position(user_side, CardPosition::Hand(user_side));
    match decision {
        MulliganDecision::Keep => {}
        MulliganDecision::Mulligan => {
            mutations::shuffle_into_deck(game, user_side, &hand)?;
            mutations::draw_cards(game, user_side, 5)?;
        }
    }

    mutations::check_start_game(game)?;

    Ok(())
}

/// The basic game action to draw a card during your turn by spending one
/// action.
#[instrument(skip(game))]
fn draw_card_action(game: &mut GameState, user_side: Side) -> Result<()> {
    verify!(
        flags::can_take_draw_card_action(game, user_side),
        "Cannot draw card for {:?}",
        user_side
    );

    debug!(?user_side, "Applying draw card action");
    mutations::spend_action_points(game, user_side, 1)?;
    let cards = mutations::draw_cards(game, user_side, 1)?;
    if let Some(card_id) = cards.get(0) {
        dispatch::invoke_event(game, DrawCardActionEvent(*card_id))?;
    }

    mutations::check_end_turn(game)?;
    Ok(())
}

/// The basic game action to play a card during your turn. Spends the resource
/// cost for a card, resolves its effects, and then moves it to the appropriate
/// new [CardPosition]. Spell, Weapon, and Artifact cards are immediately
/// revealed when played.
#[instrument(skip(game))]
fn play_card_action(
    game: &mut GameState,
    user_side: Side,
    card_id: CardId,
    target: CardTarget,
) -> Result<()> {
    verify!(
        flags::can_take_play_card_action(game, user_side, card_id, target),
        "Cannot play card {:?}",
        card_id
    );

    let card = game.card(card_id);
    debug!(?card.name, ?user_side, ?card_id, ?target, "Applying play card action");

    let definition = rules::get(card.name);
    mutations::move_card(game, card_id, CardPosition::Played(user_side, target))?;

    let actions = queries::action_cost(game, card_id);
    mutations::spend_action_points(game, user_side, actions)?;

    if flags::enters_play_face_up(game, card_id) {
        let amount = queries::mana_cost(game, card_id).with_error(|| "Card has no mana cost")?;
        mana::spend(game, user_side, ManaPurpose::PayForCard(card_id), amount)?;
        if let Some(custom_cost) = &definition.cost.custom_cost {
            (custom_cost.pay)(game, card_id)?;
        }
        game.card_mut(card_id).turn_face_up();
        game.record_update(|| GameUpdate::PlayCardFaceUp(user_side, card_id));
    }

    dispatch::invoke_event(game, CastCardEvent(CardPlayed { card_id, target }))?;
    mutations::move_card(
        game,
        card_id,
        queries::played_position(game, user_side, card_id, target)?,
    )?;

    game.history
        .push(HistoryEntry { turn: game.info.turn, event: HistoryEvent::PlayedCard(card_id) });
    mutations::check_end_turn(game)?;
    Ok(())
}

/// The basic game action to activate an ability of a card in play.
#[instrument(skip(game))]
fn activate_ability_action(
    game: &mut GameState,
    user_side: Side,
    ability_id: AbilityId,
    target: CardTarget,
) -> Result<()> {
    verify!(
        flags::can_take_activate_ability_action(game, user_side, ability_id, target),
        "Cannot activate ability {:?}",
        ability_id
    );

    game.ability_state.entry(ability_id).or_default().currently_resolving = true;
    let card = game.card(ability_id.card_id);

    debug!(?card.name, ?user_side, ?ability_id, "Applying activate ability action");

    let cost = match &rules::get(card.name).ability(ability_id.index).ability_type {
        AbilityType::Activated(cost, _) => cost,
        _ => fail!("Ability is not an activated ability"),
    };

    mutations::spend_action_points(game, user_side, cost.actions)?;
    if let Some(mana) = queries::ability_mana_cost(game, ability_id) {
        mana::spend(game, user_side, ManaPurpose::ActivateAbility(ability_id), mana)?;
    }

    if let Some(custom_cost) = &cost.custom_cost {
        (custom_cost.pay)(game, ability_id)?;
    }

    game.record_update(|| GameUpdate::AbilityActivated(user_side, ability_id));
    dispatch::invoke_event(game, ActivateAbilityEvent(AbilityActivated { ability_id, target }))?;

    game.ability_state.entry(ability_id).or_default().currently_resolving = false;
    mutations::check_end_turn(game)?;
    Ok(())
}

/// The basic game action to unveil a project card in play.
#[instrument(skip(game))]
fn unveil_action(game: &mut GameState, user_side: Side, card_id: CardId) -> Result<()> {
    verify!(
        flags::can_take_unveil_card_action(game, user_side, card_id),
        "Cannot unveil card {:?}",
        card_id
    );

    mutations::unveil_card(game, card_id)?;
    Ok(())
}

/// The basic game action to gain 1 mana during your turn by spending one
/// action.
#[instrument(skip(game))]
fn gain_mana_action(game: &mut GameState, user_side: Side) -> Result<()> {
    verify!(
        flags::can_take_gain_mana_action(game, user_side),
        "Cannot gain mana for {:?}",
        user_side
    );

    debug!(?user_side, "Applying gain mana action");
    mutations::spend_action_points(game, user_side, 1)?;
    mana::gain(game, user_side, 1);
    mutations::check_end_turn(game)?;
    Ok(())
}

fn level_up_room_action(game: &mut GameState, user_side: Side, room_id: RoomId) -> Result<()> {
    verify!(
        flags::can_take_level_up_room_action(game, user_side, room_id),
        "Cannot level up room for {:?}",
        user_side
    );
    debug!(?user_side, "Applying level up room action");
    mutations::spend_action_points(game, user_side, 1)?;
    mana::spend(game, user_side, ManaPurpose::LevelUpRoom(room_id), 1)?;
    game.record_update(|| GameUpdate::LevelUpRoom(room_id, InitiatedBy::GameAction));
    mutations::level_up_room(game, room_id)?;
    mutations::check_end_turn(game)?;
    Ok(())
}

fn spend_action_point_action(game: &mut GameState, user_side: Side) -> Result<()> {
    verify!(flags::in_main_phase(game, user_side), "Cannot spend action point for {:?}", user_side);
    mutations::spend_action_points(game, user_side, 1)?;
    mutations::check_end_turn(game)?;
    Ok(())
}

/// Handles a [PromptAction] for the `user_side` player and then removes it from
/// the queue.
fn handle_prompt_action(game: &mut GameState, user_side: Side, action: PromptAction) -> Result<()> {
    if let Some(prompt) = &game.player(user_side).card_prompt_queue.get(0) {
        verify!(
            prompt.responses.iter().any(|p| *p == action),
            "Unexpected action {:?} received",
            action
        );

        game.player_mut(user_side).card_prompt_queue.remove(0);
    } else if matches!(action, PromptAction::CardAction(_)) {
        fail!("Received action with no matching prompt {:?}", action);
    }

    match action {
        PromptAction::MulliganDecision(mulligan) => {
            handle_mulligan_decision(game, user_side, mulligan)
        }
        PromptAction::SummonAction(_)
        | PromptAction::EncounterAction(_)
        | PromptAction::AccessPhaseAction(_) => raids::handle_action(game, user_side, action),
        PromptAction::CardAction(card_action) => card_prompt::handle(game, user_side, card_action),
    }
}
