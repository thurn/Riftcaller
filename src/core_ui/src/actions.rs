// Copyright Â© Riftcaller 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashMap;
use std::fmt::Debug;

use adventure_data::adventure_action::AdventureAction;
use core_data::game_primitives::ResponseContext;
use game_data::game_actions::{GameAction, GameStateAction};
use protos::riftcaller::client_action::Action;
use protos::riftcaller::game_command::Command;
use protos::riftcaller::{ClientAction, CommandList, GameCommand, LoggingMetadata, StandardAction};
use serde_json::ser;
use user_action_data::{DebugAction, UserAction};

/// Represents an action that can be performed in the user interface. Initiating
/// a server request and performing an immediate client update are both
/// supported forms of action.
pub trait InterfaceAction {
    fn as_client_action(&self) -> Action;

    /// Converts an [InterfaceAction] into a [ClientAction].
    fn build(&self) -> ClientAction {
        ClientAction { action: Some(self.as_client_action()) }
    }
}

impl InterfaceAction for Action {
    fn as_client_action(&self) -> Action {
        self.clone()
    }
}

impl InterfaceAction for StandardAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(self.clone())
    }
}

/// Marker struct for when no action is desired.
#[derive(Debug)]
pub struct NoAction {}

impl InterfaceAction for NoAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction::default())
    }
}

impl InterfaceAction for UserAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(*self),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for DebugAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::Debug(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for GameAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::GameAction(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for GameStateAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::GameAction(GameAction::GameStateAction(*self))),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for AdventureAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::AdventureAction(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Command {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(None, vec![self.clone()])),
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Vec<Command> {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(None, self.clone())),
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Vec<Option<Command>> {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(None, self.iter().flatten().cloned().collect())),
            request_fields: HashMap::new(),
        })
    }
}

/// Returns a StandardAction to apply an optimistic UI update via a series on
/// commands on click, followed by a regular game action specified as an
/// InterfaceAction.
///
/// The provided InterfaceAction must map to a StandardAction, or else this
/// function will panic.
pub fn with_optimistic_update(
    update: Vec<Command>,
    action: impl InterfaceAction + 'static,
) -> Action {
    let payload = match action.as_client_action() {
        Action::StandardAction(standard_action) => standard_action.payload,
        _ => panic!("Expected StandardAction"),
    };

    Action::StandardAction(StandardAction {
        payload,
        update: Some(command_list(None, update)),
        request_fields: HashMap::new(),
    })
}

pub fn with_request_fields(action: impl Into<UserAction>, fields: Vec<String>) -> Action {
    Action::StandardAction(StandardAction {
        payload: payload(action.into()),
        update: None,
        request_fields: fields.into_iter().map(|name| (name, String::new())).collect(),
    })
}

pub fn payload(action: UserAction) -> Vec<u8> {
    ser::to_vec(&action).expect("Serialization failed")
}

pub fn command_list(context: Option<ResponseContext>, commands: Vec<Command>) -> CommandList {
    CommandList {
        logging_metadata: logging_metadata(context),
        commands: commands.into_iter().map(|c| GameCommand { command: Some(c) }).collect(),
        metadata: None,
    }
}

fn logging_metadata(context: Option<ResponseContext>) -> Vec<LoggingMetadata> {
    match context {
        Some(ResponseContext::Adventure(adventure_id)) => {
            vec![LoggingMetadata {
                key: "adventure_id".to_string(),
                value: adventure_id.to_string(),
            }]
        }
        Some(ResponseContext::Game(game_id)) => {
            vec![LoggingMetadata { key: "game_id".to_string(), value: game_id.to_string() }]
        }
        Some(ResponseContext::LeaveAdventure) => {
            vec![LoggingMetadata { key: "adventure_id".to_string(), value: String::new() }]
        }
        Some(ResponseContext::LeaveGame) => {
            vec![LoggingMetadata { key: "game_id".to_string(), value: String::new() }]
        }
        _ => vec![],
    }
}
