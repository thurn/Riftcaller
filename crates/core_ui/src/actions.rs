// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashMap;
use std::fmt::Debug;

use data::adventure_action::AdventureAction;
use data::game_actions::{GameAction, PromptAction};
use data::user_actions::{DebugAction, UserAction};
use protos::spelldawn::client_action::Action;
use protos::spelldawn::game_command::Command;
use protos::spelldawn::{ClientAction, CommandList, GameCommand, StandardAction};
use serde_json::ser;

/// Represents an action that can be performed in the user interface. Initiating
/// a server request and performing an immediate client update are both
/// supported forms of action.
pub trait InterfaceAction {
    fn as_client_action(&self) -> Action;

    /// Converts an [InterfaceAction] into a [ClientAction].
    fn build(&self) -> ClientAction {
        ClientAction { action: Some(self.as_client_action()) }
    }
}

impl InterfaceAction for Action {
    fn as_client_action(&self) -> Action {
        self.clone()
    }
}

impl InterfaceAction for StandardAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(self.clone())
    }
}

/// Marker struct for when no action is desired.
#[derive(Debug)]
pub struct NoAction {}

impl InterfaceAction for NoAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction::default())
    }
}

impl InterfaceAction for UserAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(*self),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for DebugAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::Debug(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for GameAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::GameAction(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for PromptAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::GameAction(GameAction::PromptAction(*self))),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for AdventureAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::AdventureAction(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Command {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(vec![self.clone()])),
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Vec<Command> {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(self.clone())),
            request_fields: HashMap::new(),
        })
    }
}

/// Returns a StandardAction to apply an optimistic UI update via a series on
/// commands on click, followed by a regular game action specified as an
/// InterfaceAction.
///
/// The provided InterfaceAction must map to a StandardAction, or else this
/// function will panic.
pub fn with_optimistic_update(
    update: Vec<Command>,
    action: impl InterfaceAction + 'static,
) -> Action {
    let payload = match action.as_client_action() {
        Action::StandardAction(standard_action) => standard_action.payload,
        _ => panic!("Expected StandardAction"),
    };

    Action::StandardAction(StandardAction {
        payload,
        update: Some(command_list(update)),
        request_fields: HashMap::new(),
    })
}

pub fn payload(action: UserAction) -> Vec<u8> {
    ser::to_vec(&action).expect("Serialization failed")
}

pub fn command_list(commands: Vec<Command>) -> CommandList {
    CommandList {
        commands: commands.into_iter().map(|c| GameCommand { command: Some(c) }).collect(),
    }
}
