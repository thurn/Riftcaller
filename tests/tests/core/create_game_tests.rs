// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use core_ui::actions::InterfaceAction;
use game_data::game::MulliganDecision;
use game_data::game_actions::{GameAction, GameStateAction};
use game_data::primitives::Side;
use insta::assert_snapshot;
use protos::spelldawn::PlayerName;
use test_utils::client_interface::HasText;
use test_utils::summarize::Summary;
use test_utils::test_session::TestSession;
use test_utils::test_session_builder::TestSessionBuilder;
use test_utils::*;
use user_action_data::{NamedDeck, NewGameAction, NewGameDebugOptions, NewGameDeck, UserAction};

static OVERLORD_DECK: NewGameDeck = NewGameDeck::NamedDeck(NamedDeck::OverlordTestSpells);
static CHAMPION_DECK: NewGameDeck = NewGameDeck::NamedDeck(NamedDeck::ChampionTestSpells);

#[test]
fn create_new_game() {
    let (game_id, overlord_id, champion_id) = test_helpers::generate_ids();
    let mut session = TestSessionBuilder::new()
        .game_id(game_id)
        .user_id(overlord_id)
        .opponent_id(champion_id)
        .do_not_connect(true)
        .build();
    let response = session.perform_action(
        UserAction::NewGame(NewGameAction {
            deck: OVERLORD_DECK,
            opponent: session.opponent_id(),
            debug_options: Some(NewGameDebugOptions {
                deterministic: true,
                ..NewGameDebugOptions::default()
            }),
            tutorial: false,
        })
        .as_client_action(),
        session.user_id(),
    );
    assert_snapshot!(Summary::run(&response));
}

#[test]
fn connect_to_new_game() {
    let (game_id, overlord_id, champion_id) = test_helpers::generate_ids();
    let mut session = TestSessionBuilder::new()
        .game_id(game_id)
        .user_id(overlord_id)
        .opponent_id(champion_id)
        .do_not_connect(true)
        .build();
    initiate_game(&mut session);

    let response = session.connect(overlord_id);
    assert!(session.user.interface.controls().has_text("Keep"));
    assert!(session.user.interface.controls().has_text("Mulligan"));
    assert_eq!(5, session.user.cards.revealed_cards().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_snapshot!(Summary::run(&response));
}

#[test]
fn mulligan_legal_actions() {
    let (game_id, overlord_id, champion_id) = test_helpers::generate_ids();
    let mut session = TestSessionBuilder::new()
        .game_id(game_id)
        .user_id(overlord_id)
        .opponent_id(champion_id)
        .do_not_connect(true)
        .build();
    initiate_game(&mut session);

    test_helpers::assert_contents_equal(
        session.legal_actions(Side::Overlord),
        vec![
            GameAction::GameStateAction(GameStateAction::MulliganDecision(
                MulliganDecision::Mulligan,
            )),
            GameAction::GameStateAction(GameStateAction::MulliganDecision(MulliganDecision::Keep)),
        ],
    );
    test_helpers::assert_contents_equal(
        session.legal_actions(Side::Champion),
        vec![
            GameAction::GameStateAction(GameStateAction::MulliganDecision(
                MulliganDecision::Mulligan,
            )),
            GameAction::GameStateAction(GameStateAction::MulliganDecision(MulliganDecision::Keep)),
        ],
    );

    session.click_on(overlord_id, "Keep");
    assert!(session.legal_actions_result(Side::Overlord).is_err());
    test_helpers::assert_contents_equal(
        session.legal_actions(Side::Champion),
        vec![
            GameAction::GameStateAction(GameStateAction::MulliganDecision(
                MulliganDecision::Mulligan,
            )),
            GameAction::GameStateAction(GameStateAction::MulliganDecision(MulliganDecision::Keep)),
        ],
    );
}

#[test]
fn keep_opening_hand() {
    let (game_id, overlord_id, champion_id) = test_helpers::generate_ids();
    let mut session = TestSessionBuilder::new()
        .game_id(game_id)
        .user_id(overlord_id)
        .opponent_id(champion_id)
        .do_not_connect(true)
        .build();
    initiate_game(&mut session);

    let response = session.click_on(overlord_id, "Keep");
    assert_eq!(0, session.user.cards.revealed_cards().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_eq!(0, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.revealed_cards().len());

    assert_snapshot!(Summary::summarize(&response));
}

#[test]
fn mulligan_opening_hand() {
    let (game_id, overlord_id, champion_id) = test_helpers::generate_ids();
    let mut session = TestSessionBuilder::new()
        .game_id(game_id)
        .user_id(overlord_id)
        .opponent_id(champion_id)
        .do_not_connect(true)
        .build();
    initiate_game(&mut session);

    let response = session.click_on(overlord_id, "Mulligan");
    assert_snapshot!(Summary::summarize(&response));

    assert_eq!(0, session.user.cards.revealed_cards().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_eq!(0, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.revealed_cards().len());
}

#[test]
fn both_keep_opening_hands() {
    let (game_id, overlord_id, champion_id) = test_helpers::generate_ids();
    let mut session = TestSessionBuilder::new()
        .game_id(game_id)
        .user_id(overlord_id)
        .opponent_id(champion_id)
        .do_not_connect(true)
        .build();
    initiate_game(&mut session);

    session.click_on(overlord_id, "Keep");
    let response = session.click_on(champion_id, "Keep");
    assert_snapshot!(Summary::summarize(&response));

    assert_eq!(5, session.user.this_player.mana());
    assert_eq!(5, session.user.other_player.mana());
    assert_eq!(5, session.opponent.this_player.mana());
    assert_eq!(5, session.opponent.other_player.mana());

    assert_eq!(3, session.user.this_player.actions());
    assert_eq!(0, session.user.other_player.actions());
    assert_eq!(0, session.opponent.this_player.actions());
    assert_eq!(3, session.opponent.other_player.actions());

    assert_eq!(6, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(6, session.opponent.cards.hand(PlayerName::Opponent).len());

    assert!(session.dusk());
}

fn initiate_game(session: &mut TestSession) {
    session.perform(
        UserAction::NewGame(NewGameAction {
            deck: CHAMPION_DECK,
            opponent: session.user_id(),
            debug_options: Some(NewGameDebugOptions {
                deterministic: true,
                ..NewGameDebugOptions::default()
            }),
            tutorial: false,
        })
        .as_client_action(),
        session.opponent_id(),
    );
    session.perform(
        UserAction::NewGame(NewGameAction {
            deck: OVERLORD_DECK,
            opponent: session.opponent_id(),
            debug_options: Some(NewGameDebugOptions {
                deterministic: true,
                ..NewGameDebugOptions::default()
            }),
            tutorial: false,
        })
        .as_client_action(),
        session.user_id(),
    );

    session.connect(session.user_id()).unwrap();
    session.connect(session.opponent_id()).unwrap();
}
