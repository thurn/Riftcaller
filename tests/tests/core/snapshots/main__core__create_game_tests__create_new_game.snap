---
source: tests/tests/core/create_game_tests.rs
assertion_line: 51
expression: "Summary::run(&response)"
---

command_list: 
    RenderScreenOverlay: "<ScreenOverlay>"
